### PR TITLE
Optimize `InMemoryDataset.resize` to avoid quadratic performance

### LINKED ISSUE #325: resize is very expensive ("quadratic" performance) (PyInf#12702)
The `.resize()` call is expensive in `versioned-hdf5` because it has "quadratic" performance. A `resize` call
1. creates a complete `data_dict` for the existing dataset and then iterates over each of the chunks
2. for each new chunk in the iteration which partially overlaps with an existing chunk it then calls `InMemoryDatasetID.can_read_direct` (in `InMemoryDataset.__getitem__`) which _iterates over all chunks again_

Here's a simple reproducer which writes 100 versions of all zero datasets of increasing size:
```
import shutil

import tempfile

import h5py
import numpy as np
from deshaw.constants.dirs import DIR_cluster_tmp

from versioned_hdf5 import VersionedHDF5File

if __name__ == '__main__':
    # create temp dir path to be cleaned up on exit
    dir = tempfile.mkdtemp(dir=DIR_cluster_tmp())
    path = f'{dir}/data.h5'
    try:
        i = 200
        j = 50_000

        with h5py.File(path, 'w') as f:
            vf = VersionedHDF5File(f)
            with vf.stage_version('r0') as sv:
                sv.create_dataset('values', data=np.zeros((i, j)), chunks=(10, 100))

            for v in range(1, 101):
                with vf.stage_version(f'r{v}') as sv:
                    # every version we grow in both i and j direction
                    new_i = i + 1
                    new_j = j + 500

                    # grow dataset and fill new entries with zeros
                    dataset = sv['values']
                    dataset.resize((new_i, new_j))
                    dataset[i:new_i, :j] = 0
                    dataset[:i, j:new_j] = 0
                    dataset[i:new_i, j:new_j] = 0

                i, j = new_i, new_j
    finally:
        shutil.rmtree(dir)
```
which has the following profiling results (note that I commented out the call to `_verify_new_chunk_reuse`, see https://github.com/deshaw/versioned-hdf5/issues/322, which otherwise makes this even slower...)
```
1708.535 <module>  12655_rewrite_versioned_reproducer_resize.py:1
├─ 1420.831 InMemoryDataset.resize  versioned_hdf5/wrappers.py:608
│  ├─ 1200.752 InMemoryDataset.__getitem__  versioned_hdf5/wrappers.py:653
│  │  ├─ 602.269 InMemoryDatasetID.can_read_direct  versioned_hdf5/wrappers.py:1319
│  │  │  ├─ 484.831 <genexpr>  versioned_hdf5/wrappers.py:1329
│  │  │  │  ├─ 353.254 [self]  versioned_hdf5/wrappers.py
│  │  │  │  └─ 131.576 isinstance  <built-in>
│  │  │  └─ 117.358 [self]  versioned_hdf5/wrappers.py
│  │  └─ 591.969 InMemoryDataset.__getitem__  h5py/_hl/dataset.py:749
│  │     └─ 576.887 Reader.read  <built-in>
│  ├─ 135.592 InMemoryDatasetID.data_dict  versioned_hdf5/wrappers.py:1263
│  │  ├─ 77.312 <dictcomp>  versioned_hdf5/wrappers.py:1292
│  │  │  └─ 72.398 spaceid_to_slice  versioned_hdf5/slicetools.py:6
│  │  │     ├─ 49.978 Tuple.__init__  ndindex/ndindex.py:121
│  │  │     │  └─ 47.663 Tuple._typecheck  ndindex/tuple.py:48
│  │  │     │     └─ 27.412 [self]  ndindex/tuple.py
│  │  │     └─ 21.082 [self]  versioned_hdf5/slicetools.py
│  │  └─ 32.373 <listcomp>  versioned_hdf5/wrappers.py:1282
│  │     └─ 31.429 [self]  versioned_hdf5/wrappers.py
│  └─ 57.820 ChunkSize.as_subchunks  ndindex/chunking.py:143
│     └─ 57.345 _indices  ndindex/chunking.py:260
│        ├─ 32.749 <listcomp>  ndindex/chunking.py:263
│        │  └─ 26.613 Slice.__init__  ndindex/ndindex.py:121
│        │     └─ 20.872 Slice._typecheck  ndindex/slice.py:55
│        └─ 21.002 Tuple.__init__  ndindex/ndindex.py:121
│           └─ 19.885 Tuple._typecheck  ndindex/tuple.py:48
├─ 210.745 _GeneratorContextManager.__exit__  contextlib.py:141
│  └─ 210.745 VersionedHDF5File.stage_version  versioned_hdf5/api.py:267
│     └─ 210.708 commit_version  versioned_hdf5/versions.py:71
│        └─ 188.322 create_virtual_dataset  versioned_hdf5/backend.py:415
│           ├─ 33.063 [self]  versioned_hdf5/backend.py
│           ├─ 25.689 select  h5py/_hl/selections.py:19
│           │  └─ 18.498 [self]  h5py/_hl/selections.py
│           ├─ 23.033 Tuple.__init__  ndindex/ndindex.py:121
│           │  └─ 21.667 Tuple._typecheck  ndindex/tuple.py:48
│           ├─ 19.534 <dictcomp>  versioned_hdf5/backend.py:423
│           └─ 17.415 Group.create_virtual_dataset  h5py/_hl/group.py:188
├─ 38.706 DatasetWrapper.__setitem__  versioned_hdf5/wrappers.py:1216
│  └─ 38.652 InMemoryDataset.__setitem__  versioned_hdf5/wrappers.py:729
│     └─ 26.848 Tuple.as_subindex  ndindex/tuple.py:626
└─ 37.036 _GeneratorContextManager.__enter__  contextlib.py:132
   └─ 37.035 VersionedHDF5File.stage_version  versioned_hdf5/api.py:267
      └─ 36.959 create_version_group  versioned_hdf5/versions.py:22
         └─ 36.854 Group.visititems  h5py/_hl/group.py:635
            └─ 31.568 proxy  h5py/_hl/group.py:660
               └─ 19.589 _get  versioned_hdf5/versions.py:57
```

### PR DESCRIPTION
**Pull request by peytondmurray**
This PR optimizes `InMemoryDataset.resize` to avoid quadratic performance with the number of chunks in `InMemoryDataset.data_dict`; closes #325.

When calling `resize()`, any chunk of the dataset which is at the edge of the new dataset's shape needs to be read, and these chunks then need to be written to the raw dataset. In the process of reading the data for each edge-chunk in `Dataset.data_dict`, each call to `InMemoryDataset.__getitem__` would execute `InMemoryDatasetID.can_read_direct`, which itself iterates through every item in `Dataset.data_dict`, leading to very slow performance for certain datasets with large `data_dict`.

This PR introduces a new `Dataset.get_index` method which takes a `can_read_direct` parameter, allowing `InMemoryDataset.resize` to call `InMemoryDatasetID.can_read_direct` just once per resize, effectively eliminating this overhead. The existing `InMemoryDataset.__getitem__` method itself calls `InMemoryDataset.get_index`.

This PR also adds a test to ensure that `resize` only results in one call to `can_read_direct` to ensure this doesn't happen again. Docstrings and type annotations were also improved.

### CODE REVIEW

File: versioned_hdf5/wrappers.py [hunk 1]
```diff
@@ -726,6 +749,18 @@ def __getitem__(self, args, new_dtype=None):
         # Return arr as a scalar if it is shape () (matching h5py)
         return arr[()]
 
+    @with_phil
+    def __getitem__(
+        self,
+        args: Union[slice, Slice, Tuple, tuple, h5r.RegionReference],
+        new_dtype: Optional[str] = None,
```
Comment (ArvidJB) [inline versioned_hdf5/wrappers.py:756]: Is anyone using `new_dtype`? Currently it does not seem to work:
```
In [1]: with TempDirCtx(DIR_cluster_tmp()) as d:
   ...:     with h5py.File(d/'foo.h5', 'w') as f:
   ...:         vf = VersionedHDF5File(f)
   ...:         with vf.stage_version('r0') as sv:
   ...:             sv.create_dataset('values', data=np.arange(10))
   ...:             values = sv['values']
   ...:             print(values[::2, np.dtype('double')])
   ...:
[PYFLYBY] from deshaw.constants.dirs import DIR_cluster_tmp
[PYFLYBY] from deshaw.file.filename import TempDirCtx
[PYFLYBY] from versioned_hdf5 import VersionedHDF5File
[PYFLYBY] import h5py
[PYFLYBY] import numpy as np
---------------------------------------------------------------------------
IndexError                                Traceback (most recent call last)
Cell In[1], line 7
      5 sv.create_dataset('values', data=np.arange(10))
      6 values = sv['values']
----> 7 print(values[::2, np.dtype('double')])

File /usr/local/python/python-3.11/std/lib64/python3.11/site-packages/versioned_hdf5/wrappers.py:950, in InMemoryArrayDataset.__getitem__(self, item)
    949 def __getitem__(self, item):
--> 950     return self.array.__getitem__(item)

IndexError: only integers, slices (`:`), ellipsis (`...`), numpy.newaxis (`None`) and integer or boolean arrays are valid indices

In [2]: with TempDirCtx(DIR_cluster_tmp()) as d:
   ...:     with h5py.File(d/'foo.h5', 'w') as f:
   ...:         vf = VersionedHDF5File(f)
   ...:         with vf.stage_version('r0') as sv:
   ...:             sv.create_dataset('values', data=np.arange(10))
   ...:         with vf.stage_version('r1') as sv:
   ...:             values = sv['values']
   ...:             print(values[::2, np.dtype('double')])
   ...:
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
Cell In[2], line 8
      6 with vf.stage_version('r1') as sv:
      7     values = sv['values']
----> 8     print(values[::2, np.dtype('double')])

File /usr/local/python/python-3.11/std/lib64/python3.11/site-packages/versioned_hdf5/wrappers.py:1129, in DatasetWrapper.__getitem__(self, index)
   1128 def __getitem__(self, index):
-> 1129     return self.dataset.__getitem__(index)

File h5py/_objects.pyx:54, in h5py._objects.with_phil.wrapper()

File h5py/_objects.pyx:55, in h5py._objects.with_phil.wrapper()

File /usr/local/python/python-3.11/std/lib64/python3.11/site-packages/versioned_hdf5/wrappers.py:658, in InMemoryDataset.__getitem__(self, args, new_dtype)
    655 idx = ndindex(args).expand(self.shape)
    657 if self.id.can_read_direct:
--> 658     return super().__getitem__(idx.raw)
    660 arr = np.ndarray(idx.newshape(self.shape), new_dtype, order='C')
    662 for chunk in self.chunks.as_subchunks(idx, self.shape):

File h5py/_objects.pyx:54, in h5py._objects.with_phil.wrapper()

File h5py/_objects.pyx:55, in h5py._objects.with_phil.wrapper()

File /usr/local/python/python-3.11/std/lib64/python3.11/site-packages/h5py/_hl/dataset.py:768, in Dataset.__getitem__(self, args, new_dtype)
    766 if self._fast_read_ok and (new_dtype is None):
    767     try:
--> 768         return self._fast_reader.read(args)
    769     except TypeError:
    770         pass  # Fall back to Python read pathway below

File h5py/_selector.pyx:361, in h5py._selector.Reader.read()

File h5py/_selector.pyx:107, in h5py._selector.Selector.apply_args()

ValueError: 2 indexing arguments for 1 dimensions
```
Probably we should just drop that?

File: versioned_hdf5/wrappers.py [hunk 2]
```diff
@@ -634,30 +635,52 @@ def resize(self, size, axis=None):
...
         self.id.shape = size
 
     @with_phil
-    def __getitem__(self, args, new_dtype=None):
-        """Read a slice from the HDF5 dataset.
+    def get_index(
+        self,
+        args: Union[slice, Slice, Tuple, tuple, h5r.RegionReference],
+        new_dtype: Optional[str] = None,
+        can_read_direct: bool = True,
```
Comment (ArvidJB) [inline versioned_hdf5/wrappers.py:661]: I think the default should be `None` and if it's not passed we default to evaluating `self.id.can_read_direct`.

**Review by ArvidJB — COMMENTED** [review]
Looks good, thanks for picking this up so quickly.

I had a comment on `new_dtype`. If that turns out to be used somewhere we can address that in a separate issue.

File: versioned_hdf5/wrappers.py (hunk 1 above)
Comment (peytondmurray) [reply versioned_hdf5/wrappers.py:756]: I'd be fine dropping this - this is actually the first time I've ever seen a `__getitem__` implementation that takes more than one argument anyway.

File: versioned_hdf5/wrappers.py (hunk 2 above)
Comment (peytondmurray) [reply versioned_hdf5/wrappers.py:661]: Sounds good, I'll make the change.

**Review by ArvidJB — APPROVED** [review]
Thanks, looks good to me!

**Pull request merged**